docs/proposals/20230407-flexible-managed-k8s-endpoints.md

```diff
@@ -0,0 +1,519 @@
+- The current implementation originates from the `Cluster.spec.ControlPlaneEndpoint` field, which defines the info we need for this proposal; but in future iterations we might consider to support more addressed or more ports for each ClusterEndpoint, similarly what is implemented in the core v1 Endpoint type.
+
+#### Infra Providers API changes
+
+This proposal does not introduce any breaking changes for the provider's API.
+
+However, Infra providers MUST deprecate the `spec.controlPlaneEndpoint` field in their `<Infra>Cluster` so they can remove it in a future API version.
+
+#### Core Cluster API Controllers changes
+
+- All the controllers working with ClusterClass objects must take into account that the `infrastructure.Ref` field could be omitted; most notably:
+  - The ClusterClass controller must ignore nil `infrastructure.Ref` fields while adding owner references to all the objects referenced by a ClusterClass.
+  - The Topology controller must skip the generation of the `<Infra>Cluster` objects when the `infrastructure.Ref` field in a ClusterClass is empty.
+
+- All the controllers working with Cluster objects must take into account that the `infrastructureRef` field could be omitted; most notably:
+  - The Cluster controller must use skip reconciling this external reference when the `infrastructureRef` is missing; also, the `status.InfrastructureReady` field must be automatically set to true in this case.
+
+- The Cluster controller must reconcile the new `ClusterEndpoint` CR. Please note that:
+  - The value from the `ClusterEndpoint` CRD must surface on the `spec.ControlPlaneEndpoint` field on the `Cluster` object.
+  - If both are present, the value from the `ClusterEndpoint` CRD must take precedence on the value from `<Infra>Cluster` objects still using the `spec.controlPlaneEndpoint`.
+
+- The Cluster controller must implement the temporary machinery to migrate to the new CRD existing Clusters and to deal with `<Infra>Cluster` objects still using the `spec.controlPlaneEndpoint` field as a way to communicate the ClusterAddress to "core" Cluster API controllers:
+  - If there is the `spec.ControlPlaneEndpoint` on the `Cluster` object but not a corresponding `ClusterEndpoint` CR, the CR must be created.
+
+#### Provider controller changes
+
+- All the `<Infra>Cluster` controller who are responsible to create a control plane endpoint
+  - As soon as the `spec.controlPlaneEndpoint` field in the `<Infra>Cluster` object will removed, the `<Infra>Cluster` controller must instead create a `ClusterEndpoint` CR to communicate the control plane endpoint to the Cluster API core controllers
+    - NOTE: technically it is possible to start creating the `ClusterEndpoint` CR *before* the removal of the `spec.controlPlaneEndpoint` field, because the new CR will take precedence on the value read from the field, but this is up to the infra provider maintainers.
```

Comment:
I think this contradicts the earlier statement

> Infra providers MUST deprecate the `spec.controlPlaneEndpoint` field in their `<Infra>Cluster` so they can remove it in a future API version

Reply:
Not sure if this is a `MUST` really, we can suggest, and provide a warning to users/providers

Reply:
Updated the language in the earlier statement